model.py

```diff
@@ -8812,42 +8812,95 @@ class ExternalIntegration(Base):
+    # applications themselves, as well as external services such as
+    # CDNs and Elasticsearch which are directly used by the
+    # applications.
+    SYSTEM_TYPE = 'system'
+    
+    # Supported protocols for ExternalIntegrations of LICENSE_TYPE.
+    OPDS_IMPORT = u'OPDS Import'
+    OVERDRIVE = DataSource.OVERDRIVE
+    BIBLIOTHECA = DataSource.BIBLIOTHECA
+    AXIS_360 = DataSource.AXIS_360
+    ONE_CLICK = DataSource.ONECLICK
+
+    LICENSE_PROTOCOLS = [
+        OPDS_IMPORT, OVERDRIVE, BIBLIOTHECA, AXIS_360, ONE_CLICK
+    ]
+    
+    # Some LICENSE_TYPE protocols imply that the data and
+    # licenses come from a specific data source.
+    DATA_SOURCE_FOR_LICENSE_PROTOCOL = {
+        OVERDRIVE : DataSource.OVERDRIVE,
+        BIBLIOTHECA : DataSource.BIBLIOTHECA,
+        AXIS_360 : DataSource.AXIS_360,
+        ONE_CLICK : DataSource.ONECLICK
+    }
+
+    # SYSTEM_TYPE integrations. Each system may have at most one of these.
+
+    # TODO: There are three CDNs, one for cover images, one for content,
+    # one for OPDS feeds. They need separate integrations.
     CDN = Configuration.CDN_INTEGRATION
-    ELASTICSEARCH = Configuration.ELASTICSEARCH_INTEGRATION
+    # TODO: Similarly for S3.
     AMAZON_S3 = Configuration.S3_INTEGRATION
-
-    # Simplified
+    ELASTICSEARCH = Configuration.ELASTICSEARCH_INTEGRATION
     CIRCULATION_MANAGER = Configuration.CIRCULATION_MANAGER_INTEGRATION
     CONTENT_SERVER = Configuration.CONTENT_SERVER_INTEGRATION
-    METADATA_WRANGLER = Configuration.METADATA_WRANGLER_INTEGRATION
     PATRON_WEB_CLIENT = u'Patron Web Client'
-
-    # Metadata
+    ADOBE_VENDOR_ID = u'Adobe Vendor ID'
+    
+    # METADATA_TYPE integrations
     BIBBLIO = u'Bibblio'
     CONTENT_CAFE = u'Content Cafe'
     NOVELIST = Configuration.NOVELIST_INTEGRATION
     NYPL_SHADOWCAT = u'Shadowcat'
     NYT = Configuration.NYT_INTEGRATION
     STAFF_PICKS = u'Staff Picks'
-
-    # Analytics
+    METADATA_WRANGLER = Configuration.METADATA_WRANGLER_INTEGRATION
+    
+    # ANALYTICS_TYPE integrations
     GOOGLE_ANALYTICS = u'Google Analytics'
 
-    # Patron Authentication
-    ADOBE_VENDOR_ID = u'Adobe Vendor ID'
-
-    # Admin Authentication
+    # ADMIN_AUTHENTICATION_TYPE integrations
     ADMIN_AUTH_TYPE = u'admin_auth'
     GOOGLE_OAUTH = u'Google OAuth'
-    ADMIN_AUTH_PROVIDERS = [GOOGLE_OAUTH]
+
+    # List of such ADMIN_AUTHENTICATION_TYPE integrations
+    ADMIN_AUTH_PROTOCOLS = [GOOGLE_OAUTH]
 
     __tablename__ = 'externalintegrations'
     id = Column(Integer, primary_key=True)
 
-    # If this integration isn't related to a Collection, it should
-    # have a provider and type.
-    provider = Column(Unicode, nullable=True)
+    # Each integration should have a protocol (explaining how we actually
+    # get information through it) and a type (explaining what the integration
+    # is for).
+    protocol = Column(Unicode, nullable=True)
     type = Column(Unicode, nullable=True)
```

Comment:
I do like `purpose`.

I was thinking external integrations would normally be looked up by type - if you stored a one-off integration without a type/purpose you would need to keep its id somewhere else so you could find it again.